### ISSUE TITLE
REST API V2 call is shutting down server

### ISSUE THREAD

**Issue by alew3** [NVIDIA/triton-inference-server#1649]
I'm using docker image **nvcr.io/nvidia/tritonserver:20.03.1-py3**

I created a mobilenetv2 saved model to run with triton (it works under tensorflow serving).

**I was trying to call it via api-v2 with the following:**
(echo -n '{ "inputs": [{ "name" : "input_image", "shape": "-1", "datatype" : "STRING", "data" : [{"b64" : "'; base64 example-integration-lib/www/assets/cowboyhat.jpg; echo '" }] }] }') | \
curl -i -w '\nTotal: %{time_total}s\n' -H "Content-Type: application/json" -d @-  http://localhost:8000/v2/models/mobilenetv2/versions/5/infer

**The server reports an error and shutdowns:**
tritonserver: /usr/include/rapidjson/document.h:1471: rapidjson::SizeType rapidjson::GenericValue<Encoding, Allocator>::Size() const [with Encoding = rapidjson::UTF8<>; Allocator = rapidjson::MemoryPoolAllocator<>; rapidjson::SizeType = unsigned int]: Assertion `IsArray()' failed.

here is my **config.pbtxt**

name: "mobilenetv2"

platform: "tensorflow_savedmodel"

input [
{
    name: "input_image"
    data_type: TYPE_STRING
    dims: [-1]
}
]

output [
{
    name: "label"
    data_type: TYPE_STRING
    dims: [5]     
},
{
    name: "confidence"
    data_type: TYPE_FP32
    dims: [5]
}
]

version_policy: { all { }}

I can understand if I'm passing a wrong parameter, but the server shouldn't just shutdown if it can't parse my request.

**Comment by deadeyegoodwin** [comment]
We're looking to see why you didn't get a better error message. But it seems that your "data" element is likely the problem. base64 isn't supported in that element.

https://github.com/kubeflow/kfserving/blob/master/docs/predict-api/v2/required_api.md#tensor-data

If you want to send data in binary format you need to use the binary_data_extension: https://github.com/NVIDIA/triton-inference-server/blob/master/docs/protocol/extension_binary_data.md

This extension to the protocol is used by the client libraries (if you run them with verbose turned on you can see what requests they produce).

**Comment by alew3** [comment]
> We're looking to see why you didn't get a better error message. But it seems that your "data" element is likely the problem. base64 isn't supported in that element.
> 
> https://github.com/kubeflow/kfserving/blob/master/docs/predict-api/v2/required_api.md#tensor-data
> 
> If you want to send data in binary format you need to use the binary_data_extension: https://github.com/NVIDIA/triton-inference-server/blob/master/docs/protocol/extension_binary_data.md
> 
> This extension to the protocol is used by the client libraries (if you run them with verbose turned on you can see what requests they produce).

Thanks for the pointers in the right direction, I'm going to try and use the binary upload route!

**Comment by CoderHam** [comment]
@alew3 could you also check with 20.06? We have made some improvements to the server-side HTTP parsing.

**Comment by alew3** [comment]
> @alew3 could you also check with 20.06? We have made some improvements to the server-side HTTP parsing.
@CoderHam 
How do I find the image for 20.06? Couldn't find it in the NGC Containers. Also tried nvcr.io/nvidia/tritonserver:20.06-py3 but it didn work.

**Comment by alew3** [comment]
I tried building the container 20.06-v1, but it asks for the following images that don't exist and I couldn't find the source code:
- nvcr.io/nvidia/pytorch:20.06-py3
- nvcr.io/nvidia/tensorflow:20.06-tf1-py3

**Comment by CoderHam** [comment]
The 20.06 containers will release on 25th. Did sending the base64 encoded image in using the binary extension as suggested by @deadeyegoodwin not work?

**Comment by alew3** [comment]
> The 20.06 containers will release on 25th. Did sending the base64 encoded image in using the binary extension as suggested by @deadeyegoodwin not work?

No luck so far. This is how far I got to getting it to run following the binary data extension documentation using curl:

(printf '{\n"model_name" : "mobilenetv2", \n"inputs":  \n{ \n "name" : "input_image" , \n "shape": [-1], \n "datatype": "BYTES", \n "parameters" : { \n "binary_data_size" : '; \
stat --printf "%s" example-integration-lib/www/assets/cowboyhat.jpg; printf ' } \n } , \n "outputs" : [ \n  { \n "name" : "label", \n "datatype" : "STRING", \n "shape" : [5] \n  \n } , \n { "name" : "confidence", \n  "datatype" : "FP32", \n "shape": [5] \n } \n  ] }') | \
curl --trace-ascii dump \
http://localhost:8000/v2/models/mobilenetv2/versions/5/infer \
-H "Content-Type:application/octet-stream" \
-H "Inference-Header-Content-Length: 317" \
-H 'Expect:' \
-w '\nTotal: %{time_total}s\n' \
-d @- \
--data-binary @example-integration-lib/www/assets/cowboyhat.jpg

It also breaks the server with this JSON:

{
    "model_name": "mobilenetv2",
    "inputs": [
        {
            "name": "input_image",
            "shape": [
                -1
            ],
            "datatype": "BYTES",
            "parameters": {
                "binary_data_size": 183534
            }
        }
    ],
    "outputs": [
        {
            "name": "label",
            "datatype": "STRING",
            "shape": [
                5
            ]
        },
        {
            "name": "confidence",
            "datatype": "FP32",
            "shape": [
                5
            ]
        }
    ]
}

Below is a pastebin with a trace that has the request body and headers.
https://pastebin.com/nXfjRdHs

**Comment by deadeyegoodwin** [comment]
What do you mean by "breaks the server"? Do you get a segfault?
One problem you have is that you can't send an input with shape -1. You need to set the shape to be the actual shape of the input.

**Comment by alew3** [comment]
> What do you mean by "breaks the server"? Do you get a segfault?
> One problem you have is that you can't send an input with shape -1. You need to set the shape to be the actual shape of the input.
@deadeyegoodwin 
Same result as the first error. Outputs an error parsing the JSON (see below) and stops running the server docker container. 

"tritonserver: /usr/include/rapidjson/document.h:1471: rapidjson::SizeType rapidjson::GenericValue<Encoding, Allocator>::Size() const [with Encoding = rapidjson::UTF8<>; Allocator = rapidjson::MemoryPoolAllocator<>; rapidjson::SizeType = unsigned int]: Assertion `IsArray()' failed."

**Comment by cloudhan** [comment]
I also encountered this issue when testing with http v2 api. The server should be absolutely resilient to the wrong request, otherwise, DOS attack will be pretty easy.

**Comment by deadeyegoodwin** [comment]
The HTTP endpoint is significantly hardened in the 20.06 release. But we are going to test your specific failure to see how it is handled.

**Comment by CoderHam** [comment]
Thanks for bringing this to our attention @cloudhan. As @deadeyegoodwin said we have hardened the HTTP endpoint and will add testing to ensure it remains tolerant. Regarding your specific issue I can confirm it was fixed in 20.06. If the shape contains negative values the server will catch that error and report it instead of crashing. (Note that we will be making the error message more descriptive.)

Request:
```
curl --data "{\"inputs\":[{\"name\":\"INPUT0\",\"shape\":[-1,16],\"datatype\":\"INT32\",\"parameters\":{},\"data\":[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]},{\"name\":\"INPUT1\",\"shape\":[1,16],\"datatype\":\"INT32\",\"parameters\":{}, \"data\":[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]}]}" http://localhost:8000/v2/models/simple/infer -v
```
Response:
```
{"error":"attempt to access JSON non-unsigned-integer as unsigned-integer"}
```

**Comment by ltm920716** [comment]
> We're looking to see why you didn't get a better error message. But it seems that your "data" element is likely the problem. base64 isn't supported in that element.
> 
> https://github.com/kubeflow/kfserving/blob/master/docs/predict-api/v2/required_api.md#tensor-data
> 
> If you want to send data in binary format you need to use the binary_data_extension: https://github.com/NVIDIA/triton-inference-server/blob/master/docs/protocol/extension_binary_data.md
> 
> This extension to the protocol is used by the client libraries (if you run them with verbose turned on you can see what requests they produce).

hello @deadeyegoodwin :
 if I want to use this binary extension in python script, where should the input bytes data place, the red line part in bellow picture, help please, thanks!
![image](https://user-images.githubusercontent.com/12007227/211756105-c39e40b0-eac2-4def-920f-b1d2b72a50da.png)
